api-select.go

```diff
@@ -0,0 +1,589 @@
+						}
+					}
+					// knits together the payload with the previous partial one.
+					myPayload = append(partialRecordVal, myPayload...)
+					if isPartial(myPayload) {
+						// if its another partial payload we need to write current complete
+						// part and set part that needs to be stitched
+						var completePayload []byte
+						completePayload, partialRecordVal, lenOfTrim = processPartial(myPayload)
+						processEvent(completePayload, headers, streamer)
+						continue
+					}
+					// if there is no longer a partial porition we can return our payload
+					// to the client and break
+					if len(myPayload) > 0 {
+						processEvent(myPayload, headers, streamer)
+					}
+					payload = nil
+					break
+				}
+			}
+			if err != nil {
+				if streamer.err == nil {
+					streamer.err = err
+				}
+			}
+		}
+		// Ensures that the full message's CRC is correct and that the message is
+		// not corrupted
+
+		if err := checkCRC(message, crc.Sum32()); err != nil {
+			if streamer.err == nil {
+				streamer.err = err
+			}
+		}
+
+		// Processes the Event and wraps it in the appropriate class, if it is the
+		// end then returns.
+		if payload != nil || headers.messageType == "End" || headers.messageType == "Cont" {
+			if end := processEvent(payload, headers, streamer); end {
+				return
+			}
+		}
+
+	}
+}
+
+// isPartial is a function which returns whether or not a message's payload is
+// partial
+func isPartial(payload []byte) bool {
+	if len(payload) < 2 {
+		return false
+	}
+	return !(payload[len(payload)-1] == byte(10) && payload[len(payload)-2] == byte(13))
+}
+
+// processPartial is the function which helps to stitch together rows when
+// messages return partial rows
+func processPartial(payload []byte) ([]byte, []byte, int) {
+	myPosition := -1
+	for i := len(payload) - 1; i >= 1; i-- {
+		if payload[i] == byte(10) && payload[i-1] == byte(13) {
+			temp := make([]byte, (myPosition - i))
+			copy(temp, payload[(i+1):(myPosition+1)])
+			return payload[:i+1], temp, len(payload) - myPosition - 1
+		}
+		if payload[i] != byte(0) && myPosition == -1 {
+			myPosition = i
+		}
+	}
+	return payload, nil, 0
+}
+
+// processEvent is a function that takes the extracted information from the
+// headers and payload to return a particular event.
+func processEvent(payload []byte, headers *headerInfo, streamer *Events) bool {
+	switch headers.messageType {
+	case "Records":
+		myTempRecord := &RecordEvent{
+			Payload: payload,
+		}
+		streamer.Response <- myTempRecord
+	case "End":
+		myEnd := &EndEvent{}
+		streamer.Response <- myEnd
+		return true
+	case "Progress":
+		myProgress := &ProgressEvent{
+			Payload: payload,
+		}
+		streamer.Response <- myProgress
+	case "Stats":
+		myStat := &StatEvent{
+			Payload: payload,
+		}
+		streamer.Response <- myStat
+	case "Cont":
+		myCont := &ContEvent{}
+		streamer.Response <- myCont
+	case "Error":
+		streamer.err = errors.New("Error Type of " + headers.errorType + " " + headers.errorMessage)
```

Comment:
Looks we should be returning true here to end the streaming as an error happened?